README.md
liboptaudio
===========

[![builds.sr.ht status](https://builds.sr.ht/~alextee/zrythm.svg)](https://builds.sr.ht/~alextee/liboptaudio?)

liboptaudio (lib optimized audio) is a library for
optimized audio subroutines, such as applying gain
on a float buffer and mixing float buffers.

Until version 1.0 is released, the API is subject
to change.

# Usage

Use `#include <optaudio/optaudio.h>` in your C/C++
project and add a dependency to the library using
the pkg-config name `optaudio`.

See the header file for more info.

# Building

To build only:

    meson build
    ninja -C build

Installation:

    ninja -C build install

# Patches/Issues
Open an issue in
https://redmine.zrythm.org/projects/zrythm/issues

# License
liboptaudio is released under the GNU Affero GPLv3+.
See the file COPYING for more details. Some files,
where specified, are licensed under different
licenses.

----
Copyright (C) 2020 Alexandros Theodotou

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.  This file is offered as-is,
without any warranty.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

----

Copyright (C) 2020 Alexandros Theodotou

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.  This file is offered as-is,
without any warranty.


tests/test.c
/*
 * Copyright (C) 2020 Alexandros Theodotou <alex at zrythm dot org>
 *
 * This file is part of liboptaudio
 *
 * liboptaudio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * liboptaudio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with liboptaudio.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "helper.h"

#include <time.h>
#include <optaudio/optaudio.h>

#include "simd_funcs.h"

#define NUM_FRAMES 128

/** 10ms. */
#define TRIGGER_USEC 10000

/**
 * @return Number of iterations.
 */
static long
mix_buffers_with_gain_timed (
  void (* func_ptr) (float *, const float *, unsigned int, float))
{
  float out[NUM_FRAMES];
  float in[NUM_FRAMES];
  for (int i = 0; i < NUM_FRAMES; i++)
    {
      in[i] = 0.3f;
    }

  long iterations = 0, usec = 0, trigger = TRIGGER_USEC;
  clock_t before = clock();
  do {
    func_ptr (out, in, NUM_FRAMES, 2.f);

    clock_t difference = clock() - before;
    usec = difference * 1000000 / CLOCKS_PER_SEC;
    iterations++;
  } while ( usec < trigger );

  printf (
    "Time taken %ld seconds %ld milliseconds %ld microseconds "
    "(%ld iterations)\n",
    usec/1000000, usec%1000000, usec % 1000, iterations);

  return iterations;
}

int main (
  int argc, const char* argv[])
{
  if (strcmp (argv[0], "fallback"))
    {
      printf ("fallback\n");
      /*long fallback =*/
        mix_buffers_with_gain_timed (
          mix_buffers_with_gain_fallback);
    }
#ifdef HAVE_AVX
  else if (strcmp (argv[0], "avx"))
    {
      printf ("avx\n");
      /*long avx =*/
        mix_buffers_with_gain_timed (
          mix_buffers_with_gain_avx);
      /*oa_assert (avx > fallback);*/
    }
#endif
#ifdef HAVE_SSE
  else if (strcmp (argv[0], "sse"))
    {
      printf ("sse\n");
      /*long sse =*/
        mix_buffers_with_gain_timed (
          mix_buffers_with_gain_sse);
      /*oa_assert (sse > fallback);*/
    }
#endif

  /*oa_assert_cmpfloat_with_epsilon (out[0], 0.6f, 0.0001f);*/

  return 0;
}


tests/helper.h
/*
 * Copyright (C) 2019 Alexandros Theodotou <alex at zrythm dot org>

 * This file is part of libaudec

 * libaudec is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * libaudec is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with libaudec.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef __AD_TESTS_HELPER_H__
#define __AD_TESTS_HELPER_H__

#include <math.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define oa_printf(a,...) \
  fprintf (stderr , a "\n", ##__VA_ARGS__)

#ifdef _WOE32
#define oa_assert(x) \
  if (!(x)) \
    { \
      oa_printf("Assertion failed: %s", #x); \
      exit(1); \
    }
#else
#define oa_assert(x) \
  if (!(x)) \
    { \
      oa_printf("Assertion failed: %s", #x); \
      raise (SIGTRAP); \
      exit(1); \
    }
#endif
#define oa_assert_cmpfloat_with_epsilon(n1,n2,epsilon) \
  oa_assert (fabsf (n1 - n2) < epsilon)
#define oa_assert_cmpfloat(n1,n2) \
  oa_assert (fabsf (n1 - n2) < FLT_EPSILON)

static inline int
str_endswith (
  const char *s, const char *t)
{
  size_t ls = strlen(s); // find length of s
  size_t lt = strlen(t); // find length of t
  if (ls >= lt)  // check if t can fit in s
  {
      // point s to where t should start and compare the strings from there
      return (0 == memcmp(t, s + (ls - lt), lt));
  }
  return 0; // t was longer than s
}

#endif


src/fallbacks.c
/*
 * Copyright (C) 2020 Alexandros Theodotou <alex at zrythm dot org>
 *
 * This file is part of liboptaudio
 *
 * liboptaudio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * liboptaudio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with liboptaudio.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "optaudio/optaudio.h"

void
mix_buffers_with_gain_fallback (
  float *      dest,
  float *      src,
  unsigned int nframes,
  float        gain)
{
  for (unsigned int i = 0; i < nframes; i++)
    {
      dest[i] += src[i] * gain;
    }
  return;
}


inc/simd_funcs.h
/*
 * Copyright (C) 2020 Alexandros Theodotou <alex at zrythm dot org>
 *
 * This file is part of liboptaudio
 *
 * liboptaudio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * liboptaudio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with liboptaudio.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef __OPTAUDIO_SIMD_FUNCS_H__
#define __OPTAUDIO_SIMD_FUNCS_H__

#ifdef __cplusplus
extern "C" {
#endif

#include "optaudio-config.h"

void
mix_buffers_with_gain_fallback (
  float *       dest,
  const float * src,
  unsigned int  nframes,
  float         gain);

#ifdef HAVE_AVX
void mix_buffers_with_gain_avx (
  float *       dest,
  const float * src,
  unsigned int  nframes,
  float         gain);
#endif

#ifdef HAVE_SSE
void mix_buffers_with_gain_sse (
  float *       dest,
  const float * src,
  unsigned int  nframes,
  float         gain);
#endif

#ifdef HAVE_SSE2
#endif

#ifdef HAVE_SSE3
#endif

#ifdef HAVE_SSSE3
#endif

#ifdef HAVE_SSE41
#endif

#ifdef HAVE_SSE42
#endif

#ifdef __cplusplus
}  /* extern "C" */
#endif

#endif /* __OPTAUDIO_SIMD_FUNCS_H__ */


inc/optaudio/optaudio.h
/*
 * Copyright (C) 2020 Alexandros Theodotou <alex at zrythm dot org>
 *
 * This file is part of liboptaudio
 *
 * liboptaudio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * liboptaudio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with liboptaudio.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifndef __OPTAUDIO_OPTAUDIO_H__
#define __OPTAUDIO_OPTAUDIO_H__

#ifdef __cplusplus
extern "C" {
#endif

/**
 * Optimized routine for mixing buffer with gain.
 *
 * This function may choose SSE over AVX if the pointers are
 * aligned to 16 byte boundary instead of 32 byte boundary to
 * reduce time to process.
 *
 * @param[in,out] dst Pointer to destination buffer, which
 *   gets updated.
 * @param[in] src Pointer to source buffer (not updated).
 * @param nframes Number of samples to process.
 * @param gain Gain to apply.
 */
void
optaudio_mix_buffers_with_gain (
  float *       dest,
  const float * src,
  unsigned int  nframes,
  float         gain);

#ifdef __cplusplus
}  /* extern "C" */
#endif

#endif /* __OPTAUDIO_OPTAUDIO_H__ */


src/simd_funcs.c
/*
 * Copyright (C) 2020 Alexandros Theodotou <alex at zrythm dot org>
 *
 * This file is part of liboptaudio
 *
 * liboptaudio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * liboptaudio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with liboptaudio.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <stdio.h>

#include "optaudio/optaudio.h"
#include "simd_funcs.h"

/**
 * Optimized routine for mixing buffer with gain.
 *
 * This function may choose SSE over AVX if the pointers are
 * aligned to 16 byte boundary instead of 32 byte boundary to
 * reduce time to process.
 *
 * @param[in,out] dst Pointer to destination buffer, which
 *   gets updated.
 * @param[in] src Pointer to source buffer (not updated).
 * @param nframes Number of samples to process.
 * @param gain Gain to apply.
 */
void
optaudio_mix_buffers_with_gain (
  float *       dest,
  const float * src,
  unsigned int  nframes,
  float         gain)
{
#ifdef HAVE_AVX
  mix_buffers_with_gain_avx (dest, src, nframes, gain);
#elif defined (HAVE_SSE)
  mix_buffers_with_gain_sse (dest, src, nframes, gain);
#else
  mix_buffers_with_gain_fallback (dest, src, nframes, gain);
#endif
}
